File: src/main/java/ca/yyx/hu/activities/UsbAttachedActivity.java
package ca.yyx.hu.activities;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.hardware.usb.UsbDevice;
import android.hardware.usb.UsbManager;
import android.os.Bundle;
import android.widget.Toast;

import ca.yyx.hu.App;
import ca.yyx.hu.aap.AapService;
import ca.yyx.hu.connection.UsbDeviceCompat;
import ca.yyx.hu.connection.UsbModeSwitch;
import ca.yyx.hu.utils.AppLog;
import ca.yyx.hu.utils.LocalIntent;
import ca.yyx.hu.utils.Settings;

/**
 * @author algavris
 * @date 30/05/2016.
 */
public class UsbAttachedActivity extends Activity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        AppLog.i("USB Intent: " + getIntent());

        UsbDevice device = LocalIntent.deviceFromIntent(getIntent());
        if (device == null) {
            AppLog.e("No USB device");
            finish();
            return;
        }

        if (App.get(this).transport().isAlive()) {
            AppLog.e("Thread already running");
            finish();
            return;
        }

        if (UsbDeviceCompat.isInAccessoryMode(device)) {
            AppLog.e("Usb in accessory mode");
            startService(AapService.createIntent(device, this));
            finish();
            return;
        }

        UsbDeviceCompat deviceCompat = new UsbDeviceCompat(device);
        Settings settings = new Settings(this);
        if (!settings.isConnectingDevice(deviceCompat)) {
            AppLog.i("Skipping device " + deviceCompat.getUniqueName());
            finish();
            return;
        }

        UsbManager usbManager = (UsbManager) getSystemService(Context.USB_SERVICE);
        UsbModeSwitch usbMode = new UsbModeSwitch(usbManager);
        AppLog.i("Switching USB device to accessory mode " + deviceCompat.getUniqueName());
        Toast.makeText(this, "Switching USB device to accessory mode " + deviceCompat.getUniqueName(), Toast.LENGTH_SHORT).show();
        if (usbMode.switchMode(device)) {
            Toast.makeText(this, "Success", Toast.LENGTH_SHORT).show();
        } else {
            Toast.makeText(this, "Failed", Toast.LENGTH_SHORT).show();
        }

        finish();
    }

    @Override
    protected void onNewIntent(Intent intent) {
        super.onNewIntent(intent);

        UsbDevice device = LocalIntent.deviceFromIntent(getIntent());
        if (device == null) {
            AppLog.e("No USB device");
            finish();
            return;
        }

        AppLog.i(UsbDeviceCompat.getUniqueName(device));

        if (!App.get(this).transport().isAlive()) {
            if (UsbDeviceCompat.isInAccessoryMode(device)) {
                AppLog.e("Usb in accessory mode");
                startService(AapService.createIntent(device, this));
            }
        } else {
            AppLog.e("Thread already running");
        }

        finish();
    }
}

File: app/src/main/java/info/anodsplace/headunit/app/UsbAttachedActivity.kt
package info.anodsplace.headunit.app

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.hardware.usb.UsbManager
import android.os.Bundle
import android.widget.Toast

import info.anodsplace.headunit.App
import info.anodsplace.headunit.aap.AapService
import info.anodsplace.headunit.connection.UsbAccessoryMode
import info.anodsplace.headunit.connection.isInAccessoryMode
import info.anodsplace.headunit.connection.uniqueName
import info.anodsplace.headunit.utils.AppLog
import info.anodsplace.headunit.utils.Settings
import info.anodsplace.headunit.utils.usbDevice

class UsbAttachedActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        AppLog.i { "USB Intent: $intent" }

        val device = intent.usbDevice
        if (device == null) {
            AppLog.e { "No USB device" }
            finish()
            return
        }

        if (App.provide(this).transport.isAlive) {
            AppLog.e { "Thread already running" }
            finish()
            return
        }

        if (device.isInAccessoryMode) {
            AppLog.e { "Usb in accessory mode" }
            startService(AapService.createIntent(device, this))
            finish()
            return
        }

        val usbManager = getSystemService(Context.USB_SERVICE) as UsbManager
        val usbMode = UsbAccessoryMode(usbManager)
        AppLog.i { "Switching USB device to accessory mode ${device.uniqueName}" }
        if (!usbMode.connectAndSwitch(device)) {
            Toast.makeText(this, "Failed to connect", Toast.LENGTH_SHORT).show()
        }

        finish()
    }

    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)

        val device = intent.usbDevice
        if (device == null) {
            AppLog.e { "No USB device" }
            finish()
            return
        }

        AppLog.i { device.uniqueName }

        if (!App.provide(this).transport.isAlive) {
            if (device.isInAccessoryMode) {
                AppLog.e { "Usb in accessory mode" }
                startService(AapService.createIntent(device, this))
            }
        } else {
            AppLog.e { "Thread already running" }
        }

        finish()
    }
}
